pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java

```diff
@@ -534,43 +527,58 @@ private static long[] toLongArray(Object value) {
       }
     }
 
-    private static int[] toIntArray(Object value) {
-      if (value instanceof int[]) {
-        return (int[]) value;
-      } else if (value instanceof IntArrayList) {
-        return ((IntArrayList) value).elements();
-      } else {
-        long[] longValues = (long[]) value;
-        int length = longValues.length;
-        int[] intValues = new int[length];
-        for (int i = 0; i < length; i++) {
-          intValues[i] = (int) longValues[i];
-        }
-        return intValues;
+    private static boolean[] toBooleanArray(int[] intArray) {
+      int length = intArray.length;
+      boolean[] booleanArray = new boolean[length];
+      for (int i = 0; i < length; i++) {
+        booleanArray[i] = intArray[i] == 1;
       }
+      return booleanArray;
     }
 
-    private static boolean[] toBooleanArray(Object value) {
-      int[] ints = (int[]) value;
-      boolean[] booleans = new boolean[ints.length];
-      for (int i = 0; i < ints.length; i++) {
-        booleans[i] = ints[i] == 1;
+    private static int[] fromBooleanArray(boolean[] booleanArray) {
```

Comment:
here, for example, we are wasting a lot of memory in case the array is not small. This is only used for multivalued booleans? If that I guess it is not that bad because it shouldn't be used that often

Reply:
Correct. These conversion happens only when we need to switch between internal type and external type. During query time, that can happen during scalar function invocation and final result rendering.